### PR TITLE
refactor(client): consolidate react router routes and 404 page

### PR DESCRIPTION
**Pull request by gordonpn**
#

## Changes this PR introduces

- The router configs are now consolidated in the `constants.js` file, making it easier to make future modifications
- The components related to the sidebar are better organized
- The contents of the sidebar render depending on the currently logged-in user's role
- Each path in the route are lazy loaded for better initial loading times
- Add a 404 page
- Reduce bundle size of antd

## How to test this PR

- Browse around the app and witness the lazy-loading
- Mistype a URL and see the 404 page
- Log in as a general user and you should not see the Admin item in the sidebar

## Checklist

- [x] Have I renamed the title of this pull request to follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)?
- [x] I will squash and merge when my PR is approved.
- [x] I have added tests to cover my changes (if applicable).

### CODE REVIEW

**Comment by Mackbellemore** [comment]
Task linked: [CU-nazj69 Refactor routes in frontend](https://app.clickup.com/t/nazj69)

**Review by KarinKazarian — DISMISSED** [review]
Beautiful!

File: client/src/components/AppRoutes.jsx [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  } else {
+    routes.push(
+      <Route exact={route.exact} component={route.component} path={route.path} key={route.path} />
+    );
+  }
+});
+
+const AppRoutes = () => {
+  return (
+    <Suspense fallback={<Loader />}>
```
Comment (Mackbellemore) [inline client/src/components/AppRoutes.jsx:29]: this is cool

File: client/src/stores/UIStore.js [hunk 2]
```diff
@@ -7,12 +7,22 @@ class UIStore {
...
 
   toggleSidebarState = () => {
     this.sidebarState = !this.sidebarState;
   };
+
+  closeSidebar = () => {
+    this.sidebarState = false;
+  };
+
+  openSidebar = () => {
```
Comment (Mackbellemore) [inline client/src/stores/UIStore.js:23]: love the three functions to set a boolean

**Review by Mackbellemore — DISMISSED** [review]
👩‍🎤

File: client/src/stores/UIStore.js (hunk 2 above)
Comment (gordonpn) [reply client/src/stores/UIStore.js:23]: it’s called convenience

**Comment by DavidLiang01** [comment]
In the image below: 
- I added "hhh" at the end of the url, should it bring the 404 page?
- I notice that the "Logout" button is showing, is it supposed to be the "Home" button since we haven't log in yet. 
[
Screenshot 2021-03-03 095152
](url)

**Comment by DavidLiang01** [comment]
the same thing happens when we go to the 404 page, we see the "log out" button
Screenshot 2021-03-03 095710

File: client/src/constants.js [hunk 3]
```diff
@@ -26,3 +28,48 @@ export const carouselContent = [
...
+    name: 'Inventory',
+    protected: true,
+    allowedRoles: rolesAvailable,
+    path: '/inventory',
+    component: lazy(() => import('components/views/Inventory')),
+  },
+  {
+    protected: false,
+    path: '/login',
+    component: lazy(() => import('components/views/Login')),
```
Comment (DavidLiang01) [inline client/src/constants.js:57]: I think you can add the attribute "exact" so it brings to 404 page

**Comment by gordonpn** [comment]
@DavidLiang01 Nice find David, i’ll take care of it

**Comment by DavidLiang01** [comment]
Maybe you can add something like this in the NavBar.js to fix the home/login/logout button 

```
array_of_paths =[ paths as string]

array_of_paths.includes(location.pathname)
```

If the pathname is not in the array, then display the home button.

What do you think?

**Review by DavidLiang01 — APPROVED** [review]
Nice work Gordon!

**Review by Mackbellemore — APPROVED** [review]

**Review by KarinKazarian — APPROVED** [review]

**Pull request merged**